src/btree/bt_npos.c

```diff
@@ -0,0 +1,352 @@
+ *     - If `path_str_offsetp` is set, return a string representation of the page's path.
+ *     - `start` is a position within the leaf page: 0 .. 1.
+ *       * When calculating a leaf page's position, use 0.5 to get the middle of the page.
+ *       * 0 and 1 are corner cases and can lead you to an adjacent page.
+ *       * Numbers outside of 0 .. 1 range will lead you to a prev/next page.
+ */
+double
+__wt_page_npos(WT_SESSION_IMPL *session, WT_REF *ref, double start, char *path_str,
+  size_t *path_str_offsetp, size_t path_str_sz_max)
+{
+    WT_PAGE_INDEX *pindex;
+    double npos;
+    uint32_t entries, slot;
+    int unused = 1; /* WT_UNUSED(snprintf) is cooked in GCC */
+
+    npos = start;
+    if (path_str)
+        *path_str_offsetp = 0;
+
+    WT_ENTER_PAGE_INDEX(session);
+    while (!__wt_ref_is_root(ref)) {
+        slot = UINT32_MAX; /* We get this invalid value in case of error */
+        __wt_ref_index_slot(session, ref, &pindex, &slot);
+        entries = pindex->entries;
+        /*
+         * Depending on the implementation, `__wt_ref_index_slot` might return an error or `slot`
+         * outside of range. Check for `slot < entries` ensures that it's a valid number. If it's
+         * not, then just skip the adjustment: the resulting number will be wrong but still within
+         * the range of the current page. Alternatively, could assign it any "reasonable" estimate
+         * like 0.5 or 0 / 1 depending on the walk direction.
+         */
+        if (slot < entries)
+            npos = (slot + npos) / entries;
+        if (path_str)
+            WT_UNUSED(unused = __wt_snprintf_len_incr(&path_str[*path_str_offsetp],
+                        path_str_sz_max - *path_str_offsetp, path_str_offsetp,
+                        "[%" PRIu32 "/%" PRIu32 "]", slot, entries));
+        __wt_ref_ascend(session, &ref, NULL, NULL);
+    }
+    WT_LEAVE_PAGE_INDEX(session);
+
+    if (path_str) {
+        path_str[*path_str_offsetp] = 0;
+    }
+
+    return (WT_CLAMP(npos, 0.0, 1.0));
+}
+
+/*
+ * !!!
+ * __find_closest_leaf --
+ *     Find the closest suitable page according to flags.
+ *     - It should not be deleted.
+ *     - If WT_READ_CACHE is set, the page should be in memory.
+ *     - If the initial ref is to a good page, it will be returned.
+ *     - If the initial ref is null, it does nothing.
+ */
+static int
+__find_closest_leaf(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
+{
+    if (*refp == NULL || F_ISSET(*refp, WT_REF_FLAG_LEAF))
+        return (0);
+    LF_SET(WT_READ_SKIP_INTL);
+    return (__wt_tree_walk(session, refp, flags));
+}
+
+/*
+ * __page_from_npos_internal --
+ *     Go to a leaf page given its normalized position. Note that this function can return a "bad"
+ *     page (deleted, locked, etc). The caller of this function should walk the tree to find a
+ *     suitable page.
+ *
+ * NOTE: Must be called within WT_WITH_PAGE_INDEX or WT_ENTER_PAGE_INDEX
```

Comment:
Can we add an assert that we hold a valid split gen?

Reply:
That assert is a part of WT_INTL_INDEX_GET maro